smartconfig/configuration.py

```diff
@@ -0,0 +1,173 @@
+
+
+class _ConfigEntryMeta(type):
+    """
+    Metaclass used to define special ConfigEntry behaviors.
+
+    Note: Using this metaclass outside of the library is currently not supported.
+    """
+
+    def __getattribute__(cls, item: str):
+        """
+        Special attribute lookup function.
+
+        If the attribute name starts with `_`, normal lookup is done, otherwise _configuration_mapping is used.
+
+        Args:
+            item: Attribute to lookup
+
+        Raises:
+            ConfigurationKeyError: The item doesn't exist.
+        """
+        # Use the normal lookup for attribute starting with `_`
+        if item.startswith('_'):
+            return super().__getattribute__(item)
+
+        if item not in cls._configuration_mapping:
+            raise ConfigurationKeyError(f"Entry {cls._path} doesn't define any {item} entry.")
+
+        return cls._configuration_mapping[item]
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any]) -> type:
+        """Add special attribute to the new entry."""
+        configuration_mapping = {key: value for key, value in dict_.items() if not key.startswith('_')}
+        defined_entries = {
+            name
+            for name
+            in chain(
+                dict_.keys(),
+                dict_.get('__annotations__', {}).keys()
+            )
+            if not name.startswith('_')
+        }
+
+        dict_["_configuration_mapping"] = configuration_mapping
+        dict_["_defined_entries"] = defined_entries
+
+        return super().__new__(cls, name, bases, dict_)
+
+
+class ConfigEntry(metaclass=_ConfigEntryMeta):
+    """
+    Baseclass for new configuration entries.
+
+    Class attributes of the subclasses can be overwrote using YAML-like configuration files.
+    The entry will use its default values and potentially directly override them with already loaded configurations,
+    and will also be overwrote in future by newly loaded configurations.
+
+    The path used by an entry is by default the `__name__` attribute of the module it is defined in.
+
+    Attributes:
+        _path_override:
+            Can be used to override the default `_path` setting.
+        _path:
+            Path used to reference this entry in the configuration files. Shouldn't be manually set.
+        _configuration_mapping:
+            Current mapping of attributes. Shouldn't be manually set.
+        _defined_entries:
+            List of attributes being registered on the class,
+            either by giving a concrete value or a typehint. Shouldn't be manually set.
+    """
+
+    _path_override: Optional[str] = None
+
+    _path: str
+    _configuration_mapping: _EntryMapping
+    _defined_entries: List[str]
+
+    @classmethod
+    def _register_entry(cls):
+        """Set the `_path` attribute and register the entry."""
+        cls._path = cls._path_override or inspect.getmodule(cls).__name__
```

Comment:
Is it be possible for a `.__module__` attribute to be empty?

Reply:
Except for manual user tampering, it's not possible as far as I know.

Reply:
In this case as far as I can tell there's no way to end up with an empty string. If `cls._path_override` is an empty string, it will have a falsy value and the module name will be used.

Reply:
Yeah this is fine. I probably had a brain fart when reviewing this.